code/parachain/frame/pablo/src/lib.rs

```diff
@@ -696,24 +669,17 @@ pub mod pallet {
 			pool_id: Self::PoolId,
 			amounts: BTreeMap<Self::AssetId, Self::Balance>,
 		) -> Result<Self::Balance, DispatchError> {
-			let pool = Self::get_pool(pool_id)?;
-			let pool_account = Self::account_id(&pool_id);
-			#[allow(deprecated)]
-			let currency_pair = Self::pool_ordered_pair(pool_id)?;
-			ensure!(amounts.len() < 3, Error::<T>::MoreThanTwoAssetsNotYetSupported);
-			let base_amount = *amounts.get(&currency_pair.base).ok_or(Error::<T>::MissingAmount)?;
-			let quote_amount =
-				*amounts.get(&currency_pair.quote).ok_or(Error::<T>::MissingAmount)?;
-			ensure!(
-				T::Assets::reducible_balance(currency_pair.base, who, false) >= base_amount,
-				Error::<T>::NotEnoughLiquidity
-			);
-			ensure!(
-				T::Assets::reducible_balance(currency_pair.quote, who, false) >= quote_amount,
-				Error::<T>::NotEnoughLiquidity
-			);
-
-			Self::lp_for_liquidity(pool, pool_account, base_amount, quote_amount)
+			with_transaction(|| {
+				// since we're simulating this, we want to immediately roll it back no matter what
+				// the outcome is
+				TransactionOutcome::Rollback(<Self as Amm>::add_liquidity(
+					who,
+					pool_id,
+					amounts,
+					Zero::zero(),
+					false,
+				))
+			})
```

Comment:
I'm not sure it's possible to mutate the storage in RPCs, I think that you can use the base API and it won't interfere with the ledger. Must be checked but I believe already discussed this with @KaiserKarel

Reply:
afaik yeah, but now I wonder too.

Reply:
it's not possible in RPCs, no, but the simulate_*_liquidity fns are exposed outside of rpcs. IMO they shouldn't exist at all, and we should just be calling the extrinsics directly in the rpc impl, but this is fine for now

Reply:
to clarify what i mean by exposed - they aren't public facing API, but they are part of the `Amm` trait; so they can technically be called from anywhere.

Reply:
Makes sense, thanks for clarifying :)